File: py/getIdTokenWithServiceAccount.py
# getIdToken.py
# ------------------------------------------------------------------
#
# A Python script that emulates the functionality of the provided
# Node.js script to obtain a Google Cloud ID token from a service
# account key.
#
# Copyright © 2025 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import argparse
import json
import time
import sys
import os
import jwt
import requests
import traceback

def create_signed_jwt(key_data, audience):
    """Creates a signed JWT using the service account credentials."""

    now_in_seconds = int(time.time())

    # Construct the JWT claims
    claims = {
        "iss": key_data["client_email"],
        "aud": key_data["token_uri"],
        "iat": now_in_seconds,
        "exp": now_in_seconds + 60,  # Token expires in 60 seconds
        "target_audience": audience
    }

    print("JWT Payload:\n" + json.dumps(claims, indent=2))

    # Sign the JWT with the private key
    signed_jwt = jwt.encode(
        claims,
        key_data["private_key"],
        algorithm="RS256"
    )
    return signed_jwt


def redeem_jwt_for_id_token(key_data, signed_assertion):
    """Exchanges the signed JWT for a Google Cloud ID token."""

    print("\nSigned Assertion:\n" + signed_assertion)

    # Make the requets to the token URI
    response = requests.post(
        key_data["token_uri"],
        data={
            "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
            "assertion": signed_assertion
        })

    response.raise_for_status()  # Raise an exception for bad status codes
    return response.json()

def get_token_info(id_token):
    """Fetches information about the obtained ID token."""

    response = requests.get(
        "https://www.googleapis.com/oauth2/v3/tokeninfo",
        params={"id_token": id_token}
    )
    response.raise_for_status()
    return response.json()


def main():
    """Main function to parse arguments and orchestrate token retrieval."""

    parser = argparse.ArgumentParser(description="Get a Google Cloud ID token for a service account.")
    parser.add_argument(
        "--keyfile",
        required=True,
        help="Path to the service account JSON key file.")
    parser.add_argument(
        "--audience",
        required=True,
        help="The desired audience for the ID token.")

    args = parser.parse_args()

    try:
        # Load the service account key file
        with open(args.keyfile, 'r') as f:
            key_data = json.load(f)

            # 1. Create a signed JWT
            signed_assertion = create_signed_jwt(key_data, args.audience)

            # 2. Redeem the JWT for an ID token
            token_response = redeem_jwt_for_id_token(key_data, signed_assertion)
            print("\nToken Response:\n" + json.dumps(token_response, indent=2))

            # 3. Get and display information about the token
            if "id_token" in token_response:
                info = get_token_info(token_response["id_token"])
                print("\nToken Info:\n" + json.dumps(info, indent=2))

    except FileNotFoundError:
        print(f"Error: Key file not found at '{args.keyfile}'", file=sys.stderr)
    except Exception as e:
        print(f"An error occurred: {e}", file=sys.stderr)
        stack_trace_string = traceback.format_exc()
        print(f"Stack Trace:\n{stack_trace_string}")

if __name__ == "__main__":
    main()

File: node/getIdTokenWithServiceAccount.js
// getIdTokenWithServiceAccount.js
// ------------------------------------------------------------------
//
// Copyright 2019-2025 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//
// uses only modules that are builtin to node, no external dependencies.
//

/* jshint esversion:9, node:true, strict:implied */
/* global process, console, Buffer */

const crypto = require("node:crypto"),
  util = require("node:util"),
  fs = require("node:fs"),
  path = require("node:path");

const GRANT_TYPE = "urn:ietf:params:oauth:grant-type:jwt-bearer";

function logWrite() {
  console.log(util.format.apply(null, arguments) + "\n");
}

const toBase64UrlNoPadding = (s) =>
  s.replace(/=/g, "").replace(/\+/g, "-").replace(/\//g, "_");

const base64EncodeString = (theString) =>
  toBase64UrlNoPadding(Buffer.from(theString).toString("base64"));

function signJwt(header, payload, key) {
  if (!header.alg) {
    throw new Error("missing alg");
  }
  if (header.alg != "RS256") {
    throw new Error("unhandled alg: " + header.alg);
  }
  const signer = crypto.createSign("sha256");
  const signatureBase = [header, payload]
    .map((x) => base64EncodeString(JSON.stringify(x)))
    .join(".");
  signer.update(signatureBase);
  const computedSignature = toBase64UrlNoPadding(signer.sign(key, "base64"));
  return signatureBase + "." + computedSignature;
}

function getGoogleAuthJwt({ options }) {
  const keyfile = options.keyfile,
    nowInSeconds = Math.floor(Date.now() / 1000),
    jwtHeader = { alg: "RS256", typ: "JWT" },
    jwtClaims = {
      iss: keyfile.client_email,
      aud: keyfile.token_uri,
      iat: nowInSeconds,
      exp: nowInSeconds + 60,
      target_audience: options.audience,
    };
  logWrite("jwt payload: " + JSON.stringify(jwtClaims, null, 2));

  return Promise.resolve({
    options,
    assertion: signJwt(jwtHeader, jwtClaims, keyfile.private_key),
  });
}

function redeemJwtForGcpToken(ctx) {
  logWrite("assertion: " + util.format(ctx.assertion));

  const url = ctx.options.keyfile.token_uri,
    headers = {
      "content-type": "application/x-www-form-urlencoded",
    },
    method = "post",
    body = `grant_type=${GRANT_TYPE}&assertion=${ctx.assertion}`;

  logWrite(
    `\nRedemption: \n  POST ${ctx.options.keyfile.token_uri} \\\n    -d grant_type=${GRANT_TYPE} \\\n    -d assertion=${ctx.assertion}`,
  );

  return fetch(url, { method, headers, body }).then(
    async (response) => await response.json(),
  );
}

function processArgs(args) {
  let awaiting = null;
  const options = {};
  try {
    args.forEach((arg) => {
      if (awaiting) {
        if (awaiting == "--keyfile") {
          options.keyfile = arg;
          awaiting = null;
        } else if (awaiting == "--audience") {
          options.audience = arg;
          awaiting = null;
        } else {
          throw new Error(`I'm confused: ${arg}`);
        }
      } else {
        switch (arg) {
          case "--keyfile":
            if (options.keyfile) {
              throw new Error("duplicate argument: " + arg);
            }
            awaiting = arg;
            break;
          case "--audience":
            if (options.audience) {
              throw new Error("duplicate argument: " + arg);
            }
            awaiting = arg;
            break;
          case "-h":
          case "--help":
            return;
          default:
            throw new Error("unexpected argument: " + arg);
        }
      }
    });
    return options;
  } catch (exc1) {
    console.log("Exception:" + util.format(exc1));
    return;
  }
}

function usage() {
  const basename = path.basename(process.argv[1]);
  console.log(
    `usage:\n  node ${basename} --keyfile SERVICE_ACCOUNT_KEYFILE --audience DESIRED_AUDIENCE\n`,
  );
}

async function showTokenInfo(payload) {
  const response = await fetch(
      `https://www.googleapis.com/oauth2/v3/tokeninfo?id_token=${payload.id_token}`,
      {
        method: "get",
        body: null,
        headers: {},
      },
    ),
    body = await response.json();
  console.log("\ntoken info:\n" + JSON.stringify(body, null, 2));
}

function main(args) {
  try {
    const options = processArgs(args);
    if (options && options.keyfile && options.audience) {
      options.keyfile = JSON.parse(fs.readFileSync(options.keyfile, "utf8"));
      if (!options.keyfile.client_email || !options.keyfile.token_uri) {
        throw new Error("that does not look like a Service Account key file.");
      }
      getGoogleAuthJwt({ options })
        .then(redeemJwtForGcpToken)
        .then(
          (payload) => (
            console.log(
              "token response:\n" +
                JSON.stringify(payload, null, 2) +
                "\n\ntoken:\n" +
                payload.id_token,
            ),
            payload
          ),
        )
        .then((payload) => showTokenInfo(payload))
        .catch((e) => console.log(util.format(e)));
    } else {
      usage();
    }
  } catch (e) {
    console.log("Exception:" + util.format(e));
  }
}

main(process.argv.slice(2));
